craft_providers/lxd/lxd.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""LXD command-line interface helpers."""
+
+import logging
+import pathlib
+import subprocess
+from typing import Optional
+
+from craft_providers.errors import details_from_called_process_error
+
+from .errors import LXDError
+
+logger = logging.getLogger(__name__)
+
+
+class LXD:
+    """Interface to `lxd` command-line.
+
+    :param lxd_path: Path to lxd.
+    """
+
+    def __init__(self, *, lxd_path: pathlib.Path = pathlib.Path("lxd")) -> None:
+        self.lxd_path = lxd_path
+
+    def init(self, *, auto: bool = False, sudo: bool = False) -> None:
+        """Initialize LXD.
+
+        Sudo is required if user is not in lxd group.
+
+        :param auto: Use default settings.
+        :param sudo: Use sudo to invoke init.
+        """
+        if sudo:
+            cmd = ["sudo"]
+        else:
+            cmd = []
+
+        cmd += [str(self.lxd_path), "init"]
+
+        if auto:
+            cmd.append("--auto")
+
+        try:
+            subprocess.run(cmd, check=True)
+        except subprocess.CalledProcessError as error:
```

Comment:
do we have a way to raise a SudoRequired specific error?

Reply:
I don't really know of a great way to do that generically because the user may have permissions to do it if they are in the correct group (normally `lxd`).  The best I think could be done is to check `/var/snap/lxd/common/lxd/unix.socket` and see if we have access to open it, but that has a lot of assumptions.  More generically, maybe we could call `lxd waitready --timeout 1` and see if it complains about "permission denied", but that forces us to wait a second and probably has its own issues.

Practically speaking, from an application perspective, the user is granting the application permission to install the provider with sudo when using Linux.  The install() method will assume permission is granted and always use sudo.  We could skip the sudo if running as root, but that logic would probably be better handled there.

Reply:
I put the sudo check into install().